myia/infer/inferrers.py

```diff
@@ -82,22 +82,168 @@ def getattr_inferrer(node, args, unif):
     return res
 
 
+def _unary_op_inference(unary_op):
+    """Create a generic inference for builtin unary operator functions."""
+
+    def inf(node, args, unif):
+        """Generic inferrer for builtin unary operator functions."""
+        (a_node,) = args
+        a_type = yield Require(a_node)
+        a_interface = a_type.tracks.interface
+        if not hasattr(a_interface, unary_op):
+            raise TypeError(f"No {unary_op} method for type {a_interface}")
+        new_node = node.graph.apply(getattr(a_interface, unary_op), a_node)
+        yield Replace(new_node)
+        res = yield Require(new_node)
+        return res
+
+    return inference_function(inf)
+
+
+def _bin_op_inference(bin_op, bin_rop):
+    """Create a generic inference for builtin binary operator functions."""
+
+    def inf(node, args, unif):
+        """Generic inferrer for builtin binary operator functions."""
+        a_node, b_node = args
+        a_type = yield Require(a_node)
+        b_type = yield Require(b_node)
+
+        if isinstance(a_type, data.GenericBase) or isinstance(
+            b_type, data.GenericBase
+        ):
+            # If there is any generic in operands, just unify them.
+            return autils.unify(a_type, b_type)[0]
```

Comment:
Ok. If I deactivate these lines, tests with `fact` will fail in `x * fact(x - 1)`. It seems it's because of `fact(x - 1)`, whose type is still Generic at this step.

I may just add a TODO here with your remarks, while keeping current code. Is it ok, or should I mark related tests as xfail in this PR ?